File: shahzar/hev.R
library(tidyverse)

time = 365
inc = 28
inf = 7
pop = 1000 # Population size

create_hh <- function() {
  # Randomly sample household sizes such that total population is 1000 individuals.
  hh_size <- sample(x = c(3, 4, 5, 6), size = 340, replace = T)
  
  # Keep households such that total population is < 1000.
  hh_size <- hh_size[which(cumsum(hh_size) < pop)]
  
  leftover <- pop - sum(hh_size)
  if (leftover < 3) {
    hh <- 1:length(hh_size)
    sampled <- sample(hh[hh_size < 6], leftover)
    hh_size[sampled] <- hh_size[sampled] + 1
  } else {
    hh_size <- c(hh_size, leftover)
  }
  return(hh_size)
}

SEIR <- function(beta_H, beta_C, inc, inf, verbose = 0) {
  hh_size = create_hh()
  
  # Create frame for running the simulation
  # ID: ID of individual
  # SIZE: size of individual's household
  # HH: ID of individual's household
  # S: susceptibility status
  # E: exposed status
  # E_count: number of days since exposed
  # I: infectious status
  # I_count: number of days since infectious
  # R: recovered status
  # INC: incubation period
  # INF: infectious period
  data = data.frame(ID = 1:pop,
                    SIZE = rep(hh_size, times = hh_size),
                    HH = rep(1:length(hh_size), times = hh_size), 
                    S = c(0, rep(1, pop - 1)), 
                    E = c(1, rep(0, pop - 1)),
                    E_count = c(1, rep(0, pop - 1)), 
                    I = 0,
                    I_count = 0, 
                    R = 0, 
                    INC = c(round(rnorm(1, inc, 2)), rep(0, pop - 1)),
                    INF = 0)
  
  # Create frame for storing results
  # ID: ID of individual
  # SIZE: size of individual's household
  # HH: ID of individual's household
  # TYPE: the kind of infection: household (H), community (C), or both (B)
  # TIME: when the individual became infectious
  # S_num: number of susceptible people in individual's household when their infectious period begins
  # I_num: number of people in household that this individual infected over their infectious period
  results <- data[, 1:3] %>% mutate(TYPE = NA, TIME = NA, S_num = NA, I_num = 0)
  results$TYPE[1] = '0'
  
  for(t in 1:time) {
    if (verbose) {
      if (t %% 10 == 0) {
        print(t)
      }
    }
    
    # Anyone who has been infectious for as many days as their infectious
    # period is now recovered.
    recovered <- (data$INF > 0) & (data$I_count == data$INF)
    if(sum(recovered, na.rm = T) > 0) {
      data$R[recovered] <- 1
      data$I[recovered] <- 0
      data$I_count[recovered] <- 0 
    }
    
    # Anyone who has been incubating for as many days as their incubation
    # period is now infectious.
    new_inf <- (data$INC > 0) & (data$E_count == data$INC)
    num_new_inf = sum(new_inf, na.rm = T)
    if(num_new_inf > 0) {
      # Change status to newly infectious and add infectious period.
      data$I[new_inf] <- 1
      random_inf <- rnorm(num_new_inf, mean = inf, sd = 1) %>% round()
      data$INF[new_inf] <- random_inf
      
      # Remove exposure status and exposure count.
      data$E[new_inf] <- 0
      data$E_count[new_inf] <- 0 
      
      # Record time at which infectious period starts.
      results$TIME[new_inf == 1] <- t
      
      # Save the number of susceptible people in each infectious 
      # individual's household.
      S_data = data %>% group_by(HH) %>% mutate(S_tot = sum(S)) %>% select(HH, S_tot)
      results$S_num[new_inf == 1] = S_data$S_tot[new_inf == 1]
    }
    
    # I_H is the number of infections inside each household.
    # I_C is the number of infections outside each household.
    I_data <- data %>% group_by(HH) %>% 
      mutate(I_H = sum(I)) %>% 
      ungroup() %>% 
      mutate(I_C = sum(I) - I_H)
    
    # Calculate household risk and community risk.
    risk_H <- beta_H * data$S * I_data$I_H / pop
    risk_C <- beta_C * data$S * I_data$I_C / pop
    
    # Each individual is infected from their household or 
    # community independently with probabilities risk_H
    # and risk_C.
    new_inf_H <- rbinom(nrow(data), 1, risk_H)
    new_inf_C <- rbinom(nrow(data), 1, risk_C)
    
    new_exposed = (new_inf_H == 1) | (new_inf_C == 1)
    num_new_exposed = sum(new_exposed, na.rm = T)
    if (sum(new_exposed) > 0) {
      # Change status to newly exposed and add incubation period.
      data$E[new_exposed] <- 1
      random_inc <- rnorm(num_new_exposed, mean = inc, sd = 2) %>% round()
      data$INC[new_exposed] <- random_inc
      
      # Remove susceptible status.
      data$S[new_exposed] <- 0
      
      # Label community infections with C and household infections with H.
      results$TYPE[new_inf_C == 1] <- 'C'
      results$TYPE[new_inf_H == 1] <- 'H'
      
      # Get number of new infections in each household.
      rr <- results %>% filter(new_inf_H == 1) %>%
        group_by(HH) %>%
        mutate(I_tot = sum(TYPE == 'H')) %>%
        ungroup()
      rr <- unique(rr[, c('HH', 'I_tot')])
      
      # Get people with the smallest I_counts and the households with the
      # new infections.
      dd = data %>% filter(data$I == 1) %>% group_by(HH) %>% slice(which.min(I_count))
      min_IDs <- dd$ID
      new_HHs <- results[new_inf_H == 1, ]$HH
      
      # Every individual with the smallest I_count in each household with 
      # at least one new H infection gets the number of new H infections
      # added to their I_num.
      idx <- (data$ID %in% min_IDs) & (data$HH %in% new_HHs)
      
      # For debugging purposes.
      if (length(results[idx, ]$I_num) != length(rr$I_tot)) {
        return(list(data, results, new_inf_H))
      }
      
      results[idx, ]$I_num <- results[idx, ]$I_num + rr$I_tot
      
      # Label individuals with both a household and community infection with B.
      results$TYPE[(new_inf_H == 1) & (new_inf_C == 1)] <- 'B'
    }
    # Increment exposure and infectious counters.
    data$E_count[data$E == 1] <- data$E_count[data$E == 1] + 1
    data$I_count[data$I == 1] <- data$I_count[data$I == 1] + 1
  }
  return(results)
}

metrics = function(results) {
  # Incidence is the proportion of the population that became infected.
  idc = mean(!is.na(results$TYPE))
  
  # If incidence is 0, the SAR is undefined.
  sar = NA
  if (idc != 0) {
    # The SAR is the average SAR for each individual that was infectious.
    sar = mean(results$I_num / results$S_num, na.rm = T)
  }
  return(c(idc, sar))
}

score = function(obs, target) {
  # The score is the L2 distance of the observed values from the target.
  return(sum((obs - target)^2))
}

# The likelihood is calculated by first averaging the incidence and SAR over N
# simulations with the state parameters. The likelihood is the negative log
# score of the average incidence and SAR.
likelihood = function(state) {
  beta_H = state[1]
  beta_C = state[2]
  
  vals = matrix(0, N, 2)
  for (i in 1:N) {
    results = SEIR(beta_H, beta_C, inc, inf)
    vals[i, ] = metrics(results)
  }
  
  avg_vals = colMeans(vals)
  print(avg_vals)
  return(-log(score(avg_vals, target)))
}

#### Metropolis algorithm ####

# Proposal function
q = function(state) {
  beta_H = state[1]
  beta_C = state[2]
  r = c(beta_H^2, beta_C^2 / 0.0001)
  v = c(beta_H^2, beta_C^2 / 0.0001)
  
  return(pmax(rgamma(n = 2, shape = r, rate = v), 1e-3))
}

# MCMC
metropolis = function(start, num_iter) {
  chain = matrix(0, num_iter + 1, 2)
  liks = matrix(0, num_iter + 1, 2)
  
  # Initialize current state.
  curr = start
  curr_lik = likelihood(curr)
  
  # Initialize best state.
  best = curr
  best_lik = curr_lik
  for (i in 1:num_iter) {
    # Save the current state and its likelihood.
    chain[i, ] = curr
    liks[i, ] = curr_lik
    
    # Print current state and likelihood.
    paste0(i, '\t', curr, " ", signif(curr_lik, 3), '\t')
    
    # Get a proposed state and calculate its likelihood.
    prop = q(curr)
    prop_lik = likelihood(prop)
    paste0(prop, " ", signif(prop_lik, 3), '\t')
    
    # Compute the ratio of the scores of the two states
    # and flip a coin.
    r = exp(prop_lik - curr_lik)
    p = runif(1)
    paste(signif(r, 3), signif(p, 3))
    
    # Transition if the proposed state is better or
    # if the coin flip succeeds.
    if (p < r) { 
      curr = prop
      curr_lik = prop_lik
      
      # If the new likelihood is better than the
      # best we've seen so far, replace the best.
      if (curr_lik > best_lik) {
        best = curr
        best_lik = curr_lik
      }
    }
    
    # Save the chain, best state, and likelihoods
    # so far.
    save(chain, file = 'chain.Rdata')
    save(liks, file = 'liks.Rdata')
    save(best, file = 'best.Rdata')
  }
  return(list(chain, liks, best))
}

# Solve for optimal values via MCMC.
target = c(0.3, 0.25) # Target values.
N = 300 # Number of times over which to average likelihood.

#metropolis_results = metropolis(c(30, 0.12), 10)
#chain = metropolis_results[[1]]
#liks = metropolis_results[[2]]
#best = metropolis_results[[3]]
#save(chain, file = "chain.Rdata")
#save(liks, file = "liks.Rdata")
#save(best, file = "best.Rdata")

t_0 = Sys.time()
beta_H = 30
beta_C = 0.15
N = 1000
vals = matrix(0, N, 2)
for (i in 1:N) {
  results = SEIR(beta_H, beta_C, inc, inf)
  vals[i, ] = metrics(results)
  save(vals, file = 'vals.Rdata')
}
t_1 = Sys.time()
print(t_1 - t_0)

File: shahzar/hev.py
import numpy as np
np.set_printoptions(formatter={'float': lambda x: "{0:0.3f}".format(x)})
import pandas as pd
from scipy import stats

time = 365
inc = 28
inf = 7

def create_hh():
    hh_size = np.random.choice([3, 4, 5, 6], size = 340, replace = True)
    hh_size = hh_size[np.cumsum(hh_size) < 1000]
    
    leftover = 1000 - hh_size.sum()
    if leftover < 3:
        # Randomly sample leftover amount indices from hh_size, and add one
        hh = np.arange(len(hh_size))
        idx = np.random.choice(hh[hh_size < 6], size = leftover, replace = False)
        hh_size[idx] += 1
    else:
        hh_size = np.append(hh_size, leftover)
    return hh_size

def SEIR(beta_H, beta_C, inc, inf, verbose = 0):
    hh_size = create_hh()
    
    # ID: ID of individual
    # SIZE: size of individual's household
    # HH: ID of individual's household
    # S: susceptibility status
    # E: exposed status
    # E_count: number of days since exposed
    # I: infectious status
    # I_count: number of days since infectious
    # R: recovered status
    # INC: incubation period
    # INF: infectious period
    # S_num: number of susceptible people in household when infectious period begins
    # I_num: number of people that this person infected over the infectious period
    data = pd.DataFrame({'ID': range(1000),
                         'SIZE': np.repeat(hh_size, repeats = hh_size),
                         'HH': np.repeat(range(len(hh_size)), repeats = hh_size),
                         'S': np.append(0, np.ones(999)),
                         'E': np.append(1, np.zeros(999)),
                         'E_count': np.append(1, np.zeros(999)),
                         'I': np.zeros(1000),
                         'I_count': np.zeros(1000),
                         'R': np.zeros(1000),
                         'INC': np.append(np.round(stats.norm.rvs(inc, 2)), np.zeros(999)),
                         'INF': np.zeros(1000)
                        })
    
    # Create frame for storing results
    results = data.loc[:, 'ID':'HH']
    results['TYPE'] = np.nan
    results.loc[0, 'TYPE'] = '0' # index case is Type 0 
    results['TIME'] = np.nan
    results['S_num'] = np.nan
    results['I_num'] = 0
    
    for t in range(time):
        if verbose:
            if t % 10 == 0: print(t, end = ' ')
        
        # Anyone who has been infectious for as many days as their infectious
        # period is now recovered.
        recovered = (data['INF'] > 0) & (data['I_count'] == data['INF'])
        if sum(recovered) > 0:
            data.loc[recovered, 'R'] = 1
            data.loc[recovered, 'I'] = 0
            data.loc[recovered, 'I_count'] = 0

        # Anyone who has been incubating for as many days as their incubation
        # period is now infectious.
        new_inf = (data['INC'] > 0) & (data['E_count'] == data['INC'])
        num_new_inf = sum(new_inf)
        if num_new_inf > 0:
            random_inf = np.round(stats.norm.rvs(inf, 1, size = num_new_inf))
            data.loc[new_inf, 'I'] = 1
            data.loc[new_inf, 'INF'] = random_inf
            data.loc[new_inf, 'E'] = 0
            data.loc[new_inf, 'E_count'] = 0
            
            # If newly infected and there isn't already a time recorded, then update with the time t.
            results['TIME'].where(~(new_inf == 1) | ~pd.isna(results['TIME']), t, inplace = True)
            #results['TIME_I'].where(~(new_inf == 1) | ~pd.isna(results['TIME_I']), t, inplace = True)
            
            # Get the number of susceptible people in each household
            S_num = data.groupby('HH').sum()

            # Get the households of the newly infectious people
            hh_idx = data.loc[new_inf, :]['HH']
            
            # Save in results the number of susceptible people
            results.loc[new_inf, 'S_num'] = S_num.loc[hh_idx, 'S'].values
            
        # I_H is the number of infections in each household.
        # I_C is the number of infections outside a given household.
        I_H = data.groupby('HH').sum()['I']
        summary = pd.DataFrame({'I_H': I_H, 
                                'I_C': data.sum()['I'] - I_H
                               })
        # dd is a frame where each individual is assigned their household's
        # I_H and I_C numbers.
        dd = data[['HH', 'S']].copy()
        dd['I_H'] = dd.apply(lambda x: summary.loc[x['HH'], 'I_H'], axis = 1)
        dd['I_C'] = dd.apply(lambda x: summary.loc[x['HH'], 'I_C'], axis = 1)

        # Calculate household risk and community risk
        risk_H = dd['S'] * beta_H * dd['I_H'] / 1000
        risk_C = dd['S'] * beta_C * dd['I_C'] / 1000

        # Calculate new household and community infections
        new_inf_H = stats.binom.rvs(1, risk_H, size = data.shape[0])
        new_inf_C = stats.binom.rvs(1, risk_C, size = data.shape[0])
        new_exposed = (new_inf_H == 1) | (new_inf_C == 1)

        num_new_exposed = sum(new_exposed)
        if num_new_exposed > 0:
            data.loc[new_exposed, 'E'] = 1
            random_inc = np.round(stats.norm.rvs(inc, 2, size = num_new_exposed))
            data.loc[new_exposed, 'INC'] = random_inc
            data.loc[new_exposed, 'S'] = 0
            
            #print(np.sum( (new_inf_H == 1) & (~pd.isna(results['TYPE'])) ))
            # The NA checks may not be necessary, since S = 0 for anyone who already has been assigned a TYPE
            results['TYPE'].where(~(new_inf_C == 1) | ~pd.isna(results['TYPE']), 'C', inplace = True)
            results['TYPE'].where(~(new_inf_H == 1) | ~pd.isna(results['TYPE']), 'H', inplace = True)
            #results['TIME_E'].where(~(new_exposed == 1) | ~pd.isna(results['TIME_E']), t, inplace = True)
            
            # Number of new infections in each household
            rr = results.loc[new_inf_H == 1, :].groupby('HH')['TYPE'].count()
            
            # Get individuals with the smallest infectious counts
            ids = data[data['I'] == 1].groupby('HH').min()['ID']
            
            # Increase I_num by the amount of people in household that got a H-type infection
            res_idx = results['ID'].isin(ids) & (results['HH'].isin(rr.index))
            results.loc[res_idx, 'I_num'] += rr.values
            
            # After the above analysis, assign 'B' to people who were infected by both.
            results['TYPE'].where(~((new_inf_H == 1) & (new_inf_C == 1)) | ~pd.isna(results['TYPE']), 'B', inplace = True)
            
        data.loc[data['E'] == 1, 'E_count'] += 1
        data.loc[data['I'] == 1, 'I_count'] += 1

    return results

def metrics(results):
    idc = np.mean(~pd.isna(results['TYPE']))
    
    sar = np.nan
    if idc != 0:
        sar = np.mean(results['I_num'] / results['S_num'])
    
    return idc, sar

def score(results, target):
    return np.sum((np.array(metrics(results)) - target)**2)

#### Metropolis algorithm ####

# Create likelihood from the score of the state
def likelihood(state):
    beta_H, beta_C = state
    
    liks = [None] * N
    for i in range(N):
        results = SEIR(beta_H, beta_C, inc, inf)
        liks[i] = -np.log(score(results, target))
    return np.mean(liks)

# Proposal function
def q(state):
    beta_H, beta_C = state
    r = [beta_H**2, beta_C**2 / 0.0001]
    v = np.array([beta_H, beta_C / 0.0001])
    return np.maximum(stats.gamma.rvs(r, scale = 1 / v, size = 2), 1e-3)

# MCMC
def metropolis(start, num_iter):
    chain = np.zeros((num_iter + 1, 2))
    liks = np.zeros((num_iter + 1, 2))
    
    # Initialize current state.
    curr = start
    curr_lik = likelihood(curr)
    
    # Initialize best state.
    best = curr
    best_lik = curr_lik
    for i in range(num_iter):
        # Save the current state and its likelihood.
        chain[i] = curr
        liks[i] = curr_lik
        
        # Print current state and likelihood.
        print(i, '\t', curr, end = ' ')
        print("%0.3f" % curr_lik, end = '\t')
        
        # Get a proposed state and calculate its likelihood.
        prop = q(curr)
        print(prop, end = ' ')
        
        prop_lik = likelihood(prop)
        print("%0.3f" % prop_lik, end = '\t')
        
        # Compute the ratio of the scores of the two states
        # and flip a coin.
        r = np.exp(prop_lik - curr_lik)
        p = stats.uniform.rvs()
        print("%0.3f" % r, ' ', "%0.3f" % p)
        # Transition if the proposed state is better or
        # if the coin flip succeeds.
        if p < r:
            curr = prop
            curr_lik = prop_lik
            
            # If the new likelihood is better than the
            # best we've seen so far, replace the best.
            if curr_lik > best_lik:
                best = curr
                best_lik = curr_lik
        
        # Save the chain, best state, and likelihoods
        # so far.
        np.save('chain.npy', chain)
        np.save('liks.npy', liks)
        np.save('best.npy', best)
    
    chain[num_iter] = curr
    liks[num_iter] = curr_lik

    np.save('chain.npy', chain)
    np.save('liks.npy', liks)
    np.save('best.npy', best)
    return chain, liks, best

# Solve for optimal values via MCMC
target = np.array([0.3, 0.25]) # target values
N = 5 # number of times over which to average likelihood

#chain, liks, best = metropolis([30, 0.12], 1000)
#np.save('chain.npy', chain)
#np.save('liks.npy', liks)
#np.save('best.npy', best)